src/stream/src/executor/top_n_appendonly.rs

```diff
@@ -179,86 +140,37 @@ impl<S: StateStore> TopNExecutorBase for InnerAppendOnlyTopNExecutor<S> {
-            assert_eq!(op, Op::Insert);
+        let mut res_ops = Vec::with_capacity(self.cache.limit);
+        let mut res_rows = Vec::with_capacity(self.cache.limit);
+        // We have banned streaming query that doesn't have LIMIT, so it is safe to unwrap here
 
+        // apply the chunk to state table
+        for (_, row_ref) in chunk.rows() {
             let pk_row = row_ref.row_by_indices(&self.internal_key_indices);
             let ordered_pk_row = OrderedRow::new(pk_row, &self.internal_key_order_types);
             let row = row_ref.to_owned_row();
-
-            if self.managed_lower_state.total_count() < self.offset {
-                // `elem` is in the range of `[0, offset)`,
-                // we ignored it for now as it is not in the result set.
-                self.managed_lower_state
-                    .insert(ordered_pk_row, row, epoch)?;
-                continue;
-            }
-
-            // We remark that when offset is 0, every input row has nothing to do with
-            // `managed_lower_state`.
-            let element_to_compare_with_upper = if self.offset > 0
-                && &ordered_pk_row < self.managed_lower_state.top_element().unwrap().0
-            {
-                // If the new element is smaller than the largest element in [0, offset),
-                // the largest element may need to move to [offset, offset+limit).
-                let res = self
-                    .managed_lower_state
-                    .pop_top_element(epoch)
-                    .await?
-                    .unwrap();
-                self.managed_lower_state
-                    .insert(ordered_pk_row, row, epoch)?;
-                res
-            } else {
-                (ordered_pk_row, row)
-            };
-
-            if self.managed_higher_state.total_count() < num_need_to_keep {
-                self.managed_higher_state.insert(
-                    element_to_compare_with_upper.0,
-                    element_to_compare_with_upper.1.clone(),
-                    epoch,
-                )?;
-                new_ops.push(Op::Insert);
-                new_rows.push(element_to_compare_with_upper.1);
-            } else if self.managed_higher_state.top_element().unwrap().0
-                > &element_to_compare_with_upper.0
-            {
-                let element_to_pop = self
-                    .managed_higher_state
-                    .pop_top_element(epoch)
-                    .await?
-                    .unwrap();
-                new_ops.push(Op::Delete);
-                new_rows.push(element_to_pop.1);
-                new_ops.push(Op::Insert);
-                new_rows.push(element_to_compare_with_upper.1.clone());
-                self.managed_higher_state.insert(
-                    element_to_compare_with_upper.0,
-                    element_to_compare_with_upper.1,
+            self.managed_state
+                .insert(ordered_pk_row.clone(), row.clone(), epoch)?;
```

Comment:
there can be some optimization for the append-only TopN because there are only insert options. in other words, we should only maintain the N rows in the state.